administrator/components/com_joomgallery/models/categories.php

```diff
@@ -721,6 +722,9 @@ public function publish($cid, $publish = 1, $task = 'publish')
       }
     }
 
+    JPluginHelper::importPlugin('content');
+    $this->_mainframe->triggerEvent('onCategoryChangeState', array(_JOOM_OPTION.'.category', $cid, array('publish'=>$publish,'task'=>$task)));
```

Comment:
According to the function signature, shouldn't just a `publish ` be passed instead of the associative array` array ('publish' => $ publish, 'task' => $ task)`?

Reply:
The associative array is for the plugin to differ from which task the event is coming from. If there would be just one task that calls the `publish` method in the `categories` model, it would make sense to pass just a `publish` to the plugin.

But there are actually two tasks calling the `publish` method in the `categories` model.
The `approve` method in the `categories` controller is calling the `publish` method too:
https://github.com/JoomGalleryfriends/JoomGallery/blob/3d2145f717a1234b1ce9d1b91375a0c724051825/administrator/components/com_joomgallery/controllers/categories.php#L143

Therefore there are two possibilities how the `onCategoryChangeState` event could be triggered. task could be `publish` or `approve`. Just passing a `publish` would not be enough information for the plugin. Or am I wrong?

Reply:
The trigger calls
```
	/**
	 * Smart Search change category state content method.
	 * Method is called when the state of the category to which the
	 * content item belongs is changed.
	 *
	 * @param   string   $extension  The extension whose category has been updated.
	 * @param   array    $pks        A list of primary key ids of the content that has changed state.
	 * @param   integer  $value      The value of the state that the content has been changed to.
	 *
	 * @return  void
	 *
	 * @since   2.5
	 */
	public function onCategoryChangeState($extension, $pks, $value)
	{
		$dispatcher = JEventDispatcher::getInstance();
		JPluginHelper::importPlugin('finder');

		// Trigger the onFinderCategoryChangeState event.
		$dispatcher->trigger('onFinderCategoryChangeState', array($extension, $pks, $value));
	}
```
The signature says **_value_** should be an **_int_**!

Inside this function the trigger calls
```
	/**
	 * Method to update the item link information when the item category is
	 * changed. This is fired when the item category is published or unpublished
	 * from the list view.
	 * Event is triggered at the same time as the onCategoryChangeState event
	 *
	 * @param   string   $extension  The extension whose category has been updated.
	 * @param   array    $pks        A list of primary key ids of the content that has changed state.
	 * @param   integer  $value      The value of the state that the content has been changed to.
	 *
	 * @return  void
	 *
	 * @since   2.5
	 */
	public function onFinderCategoryChangeState($extension, $pks, $value)
	{
		// Make sure we're handling joomgallery categories.
		if ($extension === 'com_joomgallery')
		{
			$this->categoryStateChange($pks, $value);
		}
	}
```
The signature says **_value_** should be an **_int_**! 
By the way the function `$this->categoryStateChange()` will  not be executed because you check for `if ($extension === 'com_joomgallery')` but you pass `'com_joomgallery.category'` when triggering. 

The function called inside is
```
	/**
	 * Method to update index data on category access level changes
	 *
	 * @param   array    $pks    A list of primary key ids of the content that has changed state.
	 * @param   integer  $value  The value of the state that the content has been changed to.
	 *
	 * @return  void
	 *
	 * @since   2.5
	 */
	protected function categoryStateChange($pks, $value)
	{
		/*
		 * The item's published state is tied to the category
		 * published state so we need to look up all published states
		 * before we change anything.
		 */
		foreach ($pks as $pk)
		{
			$query = clone $this->getStateQuery();
			$query->where('c.id = ' . (int) $pk);

			// Get the published states.
			$this->db->setQuery($query);
			$items = $this->db->loadObjectList();

			// Adjust the state for each item within the category.
			foreach ($items as $item)
			{
				// Translate the state.
				$temp = $this->translateState($item->state, $value);

				// Update the item.
				$this->change($item->id, 'state', $temp);

				// Reindex the item
				$this->reindex($item->id);
			}
		}
	}
```
The signature says **_value_** should be an **_int_**!

By the way, as far as I know approving a category can not be done anywhere in the JoomGallery GUI and if called directly via URL it would throw an error because there is no 'approved' field in the category database table. The code was probably prepared at some point in the past to approve categories, but it never got that far ;-).

Reply:
If we delete the "approving" in the code I am fine with just passing the integer state (0: unpublished, 1:published), to the event. I guess this will improve the compatibility of our code to the joomla core,

Reply:
I don't understand what the removal of the prepared but unused code has to do with the incorrect parameter transfer or with the  improvement of the compatibility of our code to the joomla core . If there was the function of approving categories, you couldn't just remove that code either.

You trigger a function of a Joomla plugin and according to the signature it expects an integer value when passing the parameters. This Joomla plugin triggers the `onFinderCategoryChangeState` function of all registered finder plugins, including those in the JoomFinderPlugin you provided. Your function also expects an integer value when passing this parameter. All other functions that transfer this parameter always expect an integer value (e.g. your Override of `translateState()` in the JoomFinderPlugin - you don't use this value in the function, which I don't understand either).

I know that PHP is not type safe and that you can pass any type when passing parameters, but this is bad programming style. In addition, the code will not work, because all functions interpret the "**_p_**" from "array ('publish' => $ publish, 'task' => $ task)" as integer value for the state and this value is always constant.